exchanges/gateio/ws_ob_update_manager.go

```diff
@@ -0,0 +1,152 @@
+package gateio
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common/key"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream/buffer"
+)
+
+var (
+	errOutOfOrder              = errors.New("out of order update")
+	errOrderbookSnapshotBehind = errors.New("orderbook snapshot is behind update")
+	wsOBUpdateMgr              = &wsOBUpdateManager{m: make(map[key.PairAsset]*updateCache)}
+)
+
+type wsOBUpdateManager struct {
+	m   map[key.PairAsset]*updateCache
+	mtx sync.Mutex
+}
+
+type updateCache struct {
+	buffer   []updateWithFirstID
+	updating bool
+	mtx      sync.Mutex
+}
+
+type updateWithFirstID struct {
+	update  *orderbook.Update
+	firstID int64
+}
+
+func (m *wsOBUpdateManager) applyUpdate(ctx context.Context, g *Gateio, limit uint64, firstID int64, update *orderbook.Update) error {
+	book, err := g.Websocket.Orderbook.GetOrderbook(update.Pair, update.Asset)
+	if err != nil && !errors.Is(err, buffer.ErrDepthNotFound) {
+		return err
+	}
+
+	cache := m.getCache(update.Pair, update.Asset)
+	cache.mtx.Lock()
+	defer cache.mtx.Unlock()
+
+	if cache.updating {
+		cache.buffer = append(cache.buffer, updateWithFirstID{update: update, firstID: firstID})
+		return nil
+	}
+
+	if book == nil || book.LastUpdateID+1 < firstID /*orderbook is behind notifications so refresh*/ {
+		cache.updating = true
+		go func() {
+			// REST Orderbook IDs are behind the ws update feed by about 50-100 changes; inline rudimentary delay to cache
+			time.Sleep(time.Second)
+			if err := cache.updateOrderbookAndApply(ctx, g, update.Pair, update.Asset, limit); err != nil {
+				g.Websocket.DataHandler <- err
```

Comment:
So I am forcing an issue, but if this errors, there doesn't seem to be anything else that happens to recover.

On the next update we get to the if statement
`if book.LastUpdateID+1 > update.UpdateID {`
and the update.UpdateID is larger by a degree versus book.LastUpdateID+1, so it seems like the snapshot is behind the update and there are no checks for that if `updateOrderbookAndApply` errors.

Is this okay or am I missing things?

Reply:
Yeah I only added an error because it was in the docs; once synced up I have run this over night and it didn't happen at all. For this to trigger you say `update.UpdateID is larger by a degree versus book.LastUpdateID+1` but its actually the inverse, the updates are behind the book all of a sudden 🤷 . In the off chance it does it starts giving us old updates it will complain and then should catch up? 🤷 I could probably remove this error check altogether if you want.